include/utilities/parallel_utils.h

```diff
@@ -127,6 +131,297 @@ namespace parallel {
         return is_hydrofabric_subdivided(mpi_rank, mpi_num_procs, false);
     }
 
+    /**
+     * Set each rank's host "id" value in a provided host array.
+     *
+     * This function has the ranks communicate their hostname back to rank 0.  Rank 0 then builds an array of unique
+     * host id values.  The value at index ``x`` in this array is the id for the host of rank ``x``.
+     *
+     * The id of the host of rank 0 is ``0``, but otherwise there is no deterministic ordering of id values for hosts.
+     *
+     * This array is then broadcast by rank 0 back to to the other ranks, allowing them to set their analogous arrays.
+     * It is expected all ranks run this function at the same time.
+     *
+     * @param mpi_rank The current rank.
+     * @param mpi_num_procs The number of ranks.
+     * @param host_array A pointer to an allocated array of size ``mpi_num_procs``.
+     */
+    void get_hosts_array(int mpi_rank, int mpi_num_procs, int *host_array) {
+        // Things should never be longer than this
+        int hostnameBufferSize = 256;
+        char hostnames[mpi_num_procs][hostnameBufferSize];
+        // These are the lengths of the (trimmed) C-string representations of the hostname for each rank
+        int actualHostnameCStrLength[mpi_num_procs];
+        // Initialize to -1 to represent unknown
+        for (int i = 0; i < mpi_num_procs; ++i) {
+            actualHostnameCStrLength[i] = -1;
+        }
+
+        // Get this rank's hostname
+        gethostname(hostnames[mpi_rank], hostnameBufferSize);
+        // Set the one for this rank
+        actualHostnameCStrLength[mpi_rank] = std::strlen(hostnames[mpi_rank]);
+
+        if (mpi_rank != 0) {
+            // First, send how long the string is
+            MPI_Send(&actualHostnameCStrLength[mpi_rank], 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
```

Comment:
This might be a good place for [MPI_Gather](https://www.rookiehpc.com/mpi/docs/mpi_gather.php)